taichi/runtime/llvm/runtime.cpp

```diff
@@ -1741,6 +1744,40 @@ i32 wasm_materialize(Context *context) {
   return (i32)(size_t)context->runtime->roots[0];
 }
 
+void wasm_set_print_buffer(Context *context, Ptr buffer) {
+  context->runtime->wasm_print_buffer = buffer;
+}
+
+void wasm_print_i32(Context *context, i32 value) {
+  Ptr buffer = context->runtime->wasm_print_buffer;
+  if (buffer == nullptr)
+    return;
+  i32 total_cnt = ((i32 *)buffer)[0]++;
+  i32 print_pos = total_cnt % (1024 * 1024);
+  ((i32 *)buffer)[print_pos * 2 + 1] = 0;  // 0 for i32
+  ((i32 *)buffer)[print_pos * 2 + 2] = value;
+}
+
+void wasm_print_f32(Context *context, f32 value) {
+  Ptr buffer = context->runtime->wasm_print_buffer;
+  if (buffer == nullptr)
+    return;
+  i32 total_cnt = ((i32 *)buffer)[0]++;
+  i32 print_pos = total_cnt % (1024 * 1024);
+  ((i32 *)buffer)[print_pos * 2 + 1] = 1;  // 1 for f32
+  ((f32 *)buffer)[print_pos * 2 + 2] = value;
+}
+
+void wasm_print_char(Context *context, i32 value) {
```

Comment:
Can we pack four chars together to save some space?

Reply:
The current struct works like
```
i32 total_print_character_num;
struct {
    int type;  // 0 for i32, 1 for f32, 2 for char
    union {
      i32 i;
      f32 f;
      char c;
    } data;
} wasm_buffer_buffer[kWasmPrintBufferSize];
```

Packed mode is hard to implement here... I added the above memory layout into the comment in taichi to help understand the design here. https://github.com/taichi-dev/taichi/pull/2910/commits/0413cfd4b19382c4fbd8c1e9bc215b20e08b38e8

Reply:
It feels weird to take 8 bytes to store a single char... I think by padding the string to a multiple of 4, we can easily pack four chars? I.e.

```cpp
i32 total_print_character_num;
struct {
    int type;  // 0 for i32, 1 for f32, 2 for char
    union {
      i32 i;
      f32 f;
      i32 ch4;
    } data;
} wasm_buffer_buffer[kWasmPrintBufferSize];

```